### PR TITLE
Add relations to HIRM in an order respecting their base_relations

### PR DESCRIPTION
**Pull request by ThomasColthurst**
Fixes https://github.com/probcomp/hierarchical-irm/issues/196

### CODE REVIEW

File: cxx/hirm.cc [hunk 1]
```diff
@@ -3,16 +3,38 @@
 
 #include "hirm.hh"
 
-HIRM::HIRM(const T_schema& schema, std::mt19937* prng) {
-  // Add the clean relations before the noisy relations.
-  for (const auto& [name, relation] : schema) {
-    if (std::holds_alternative<T_clean_relation>(relation)) {
-      this->add_relation(prng, name, relation);
+HIRM::HIRM(const T_schema& _schema, std::mt19937* prng) {
+  printf("in hirm\n");
```
Comment (emilyfertig) [inline cxx/hirm.cc:7]: remove?

File: cxx/hirm.cc [hunk 2]
```diff
@@ -3,16 +3,38 @@
...
-    if (std::holds_alternative<T_noisy_relation>(relation)) {
-      this->add_relation(prng, name, relation);
+
+    if (!have_pending_relations) {
+      return;
+    }
+
+    if (!added_a_relation) {
+      printf("Detected loop in schema\n");
+      std::exit(1);
```
Comment (emilyfertig) [inline cxx/hirm.cc:37]: We've used asserts everywhere else, let's stick with that.

File: cxx/hirm.cc [hunk 3]
```diff
@@ -3,16 +3,38 @@
...
-  for (const auto& [name, relation] : schema) {
-    if (std::holds_alternative<T_noisy_relation>(relation)) {
-      this->add_relation(prng, name, relation);
+
+    if (!have_pending_relations) {
+      return;
+    }
+
+    if (!added_a_relation) {
+      printf("Detected loop in schema\n");
```
Comment (emilyfertig) [inline cxx/hirm.cc:36]: I think it'd be helpful to print the relation name.

**Review by emilyfertig — APPROVED** [review]

File: cxx/hirm.cc (hunk 1 above)
Comment (ThomasColthurst) [reply cxx/hirm.cc]: Done.

File: cxx/hirm.cc (hunk 2 above)
Comment (ThomasColthurst) [reply cxx/hirm.cc:36]: assert's are removed in optimized mode, so I've tried not to use them in places like this where continuing will lead to bad behavior (an infinite loop in this case).  grep says that there are 19 such places in the codebase.

File: cxx/hirm.cc (hunk 3 above)
Comment (ThomasColthurst) [reply cxx/hirm.cc:35]: This is outside of the loop over relations, so there is no single relation name to print.

**Pull request merged**